Add Firefox add-on security documentation (#4838)
sourcegraph/sourcegraph: doc/integration/firefox_security.md

# Sourcegraph Firefox Add-on security

## Why is Firefox flagging the Sourcegraph add-on as unsafe?

> _Firefox has determined that the following add-ons are known to cause stability or security problems_

> ![Firefox browser add-on error](../img/firefox-error.png "Firefox browser add-on error")

Reviewer: ```suggestion
![Firefox browser add-on error](img/firefox-error.png "Firefox browser add-on error")
```

The Sourcegraph Firefox add-on has been flagged as unsafe by Mozilla because of a compliance issue with Mozilla's policy regarding [add-on development practices](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/AMO/Policy/Reviews#Development_Practices). This issue is specifically related to how we have chosen to implement Sourcegraph extensions.

> We made Sourcegraph extensions centrally managed by your Sourcegraph instance, not individually managed in your Firefox profile. Our customers are companies that roll out the browser extension to all employees, and asking each employee to individually manage Sourcegraph settings in Firefox would be too complex and present more security problems than central management.

Author: ```suggestion
We made Sourcegraph extensions centrally managed by your Sourcegraph instance, not individually managed in your Firefox profile. Our customers are companies that roll out the browser extension to all employees, and asking each employee to individually manage Sourcegraph settings in Firefox would be unnecessarily complex for both users and admins than the centrally managed solution.
```

Author: @nicksnyder - how's this?

Reviewer: lgtm, thanks!

## What are Sourcegraph extensions, and who can author them?

Sourcegraph extensions provide a way to extend the functionality of both a Sourcegraph instance and our browser add-on by writing JavaScript code against the Sourcegraph extension API. Sourcegraph uses extensions to implement some core features, for example, code intelligence extensions for different languages. Sourcegraph extensions also provide an opportunity for third-party developers to extend Sourcegraph's functionality.

> In the Sourcegraph extension registry, users can always examine the JavaScript bundle of the extension, and, if provided by the author, its source repository. Extensions are always opt-in, except for trusted language extensions providing code intelligence, that are written and maintained by Sourcegraph.

Reviewer: ```suggestion
In the Sourcegraph extension registry, users can always examine the JavaScript bundle of the extension, and, if provided by the author, its source repository. Extensions are always opt-in, except for trusted language extensions providing code intelligence that are written and maintained by Sourcegraph.
```

Mozilla’s main objection to our execution model is the fact that extensions upgrade automatically without user interaction, so the add-on will always fetch the latest version of the extension from your Sourcegraph instance. In order to be compliant, we would need to change this so that users always have to manually review and approve extension updates. This is a change we are not planning to implement at this time.

## What should I do?

We know that Sourcegraph users are a very technical audience, so we hope the above information will help you make an informed decision on whether to keep using the Sourcegraph Firefox add-on. Additionally, our add-on is [fully open source](https://github.com/sourcegraph/sourcegraph/tree/master/browser), and you should feel free to inspect the source code if you have any other concerns.

If you decide that you are comfortable trusting the Sourcegraph Mozilla Firefox add-on:

*   When you see the pop up recommending you disable Sourcegraph, uncheck the disable option.

OR

> *   If you already restarted with the Sourcegraph add-on disabled, you can go to your [addons page](about:addons) and re-enable Sourcegraph

Reviewer: @sqs @ryan-blunden The build is failing because `about:addons` is an invalid link. Is there a way to whitelist this?

@christinaforney can you make this look like and absolute link?

Reviewer: ```suggestion
*   If you already restarted with the Sourcegraph add-on disabled, you can go to your addons page (i.e. type `about:addons` in the Firefox address bar) and re-enable Sourcegraph
```
